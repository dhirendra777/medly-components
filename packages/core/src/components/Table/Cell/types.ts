import { HTMLProps } from '@medly-components/utils';
import { ColumnConfig } from '../types';

export interface StyledProps {
    hide?: boolean;
    frozen?: boolean;
    align?: 'left' | 'right' | 'center';
    isRowSelectionCell?: boolean;
}

export interface Props extends StyledProps, HTMLProps<HTMLDivElement> {
    data: any;
    rowId: any;
    config: ColumnConfig;
    dottedFieldName: string;
    isLoading?: boolean;
    selectedRows?: number[];
    isRowSelected?: boolean;
    isRowClickDisabled?: boolean;
    isRowSelectionDisabled?: boolean;
    onRowSelection?: (id: number) => void;
    addColumnMaxSize: (field: string, value: number) => void;
}
