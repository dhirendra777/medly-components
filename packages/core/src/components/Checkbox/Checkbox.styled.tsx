import { CheckIcon } from '@medly-components/icons';
import { defaultTheme } from '@medly-components/theme';
import { centerAligned, styled, WithThemeProp } from '@medly-components/utils';
import { CheckboxWrapperProps, Props } from './types';

export const Wrapper = styled('div')<CheckboxWrapperProps>`
    position: relative;
    width: ${({ theme, size }) => (size ? theme.checkbox.sizes[size] : theme.checkbox.defaultSize)};
    height: ${({ theme, size }) => (size ? theme.checkbox.sizes[size] : theme.checkbox.defaultSize)};
    border: 1px solid ${({ theme }) => theme.checkbox.borderColor};
    border-radius: 15%;

    &:focus-within {
        box-shadow: 0 0 0 1pt ${({ theme }) => theme.checkbox.outlineColor};
    }

    & > * {
        width: 100%;
        height: 100%;
        border-radius: 15%;
    }

    ${centerAligned()}
`;
Wrapper.defaultProps = {
    theme: defaultTheme
};

export const CheckIconWrapper = styled('div')<WithThemeProp>`
    z-index: 1;
    position: absolute;
    background-color: ${({ theme }) => theme.checkbox.bgColor};

    ${CheckIcon.Style} {
        width: 0;
        height: 0;
        transition: width 0.2s ease, height 0.2s ease;
    }

    ${centerAligned()}
`;
CheckIconWrapper.defaultProps = {
    theme: defaultTheme
};

export const Checkbox = styled('input').attrs({ type: 'checkbox' })<Props>`
    opacity: 0;
    z-index: 2;
    position: absolute;
    margin: 0;
    cursor: pointer;

    &:focus {
        outline: none;
    }

    &:disabled {
        cursor: not-allowed;
        & + ${CheckIconWrapper} {
            background: ${({ theme }) => theme.checkbox.disabledBgColor};
        }
    }

    &:checked + ${CheckIconWrapper} {
        ${CheckIcon.Style} {
            width: calc(70%);
            height: calc(70%);
        }
    }
`;
