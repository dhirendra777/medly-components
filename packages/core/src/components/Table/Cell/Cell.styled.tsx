import { SvgIcon } from '@medly-components/icons';
import { css, styled } from '@medly-components/utils';
import Checkbox from '../../Checkbox';
import Text from '../../Text';
import { StyledProps } from './types';

export const CustomComponentWrapper = styled('div')``;

export const LoadingDiv = styled('div')`
    @keyframes placeHolderShimmer {
        0% {
            background-position: -468px 0;
        }
        100% {
            background-position: 468px 0;
        }
    }
    width: 100%;
    height: 100%;
    animation-duration: 1.25s;
    animation-fill-mode: forwards;
    animation-iteration-count: infinite;
    animation-name: placeHolderShimmer;
    animation-timing-function: linear;
    background: ${({ theme }) => theme.colors.grey[100]};
    background: linear-gradient(
        to right,
        ${({ theme }) => theme.colors.grey[100]} 8%,
        ${({ theme }) => theme.colors.grey[200]} 18%,
        ${({ theme }) => theme.colors.grey[100]} 33%
    );
    background-size: 800px 104px;
    position: relative;
`;

export const Cell = styled('div')<StyledProps>`
    opacity: ${({ hide }) => (hide ? 0 : 1)};
    padding: ${({ hide }) => (hide ? '0' : '5px 10px')};
    position: relative;
    display: flex;
    justify-content: ${({ align }) => (align === 'right' ? 'flex-end' : align === 'center' ? 'center' : 'flex-start')};
    align-items: center;
    overflow: hidden;
    border-right: 1px solid ${({ theme }) => theme.table.borderColor};
    cursor: ${({ isRowSelectionCell }) => (isRowSelectionCell ? 'default' : 'inherit')};

    ${Checkbox.Style} {
        display: flex;
        margin: 0px;
        ${SvgIcon} {
            margin-right: 0px;
        }
    }

    ${Text.Style} {
        white-space: nowrap;
    }

    ${props =>
        props.frozen &&
        css`
            position: sticky;
            left: 0;
            z-index: 2;
            background-color: inherit;

            * {
                z-index: 2;
            }
        `}
`;
