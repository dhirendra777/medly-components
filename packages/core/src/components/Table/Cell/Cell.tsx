import { WithStyle } from '@medly-components/utils';
import React, { useCallback, useEffect, useMemo, useRef } from 'react';
import Checkbox from '../../Checkbox';
import Text from '../../Text';
import * as Styled from './Cell.styled';
import { Props } from './types';

const Cell: React.SFC<Props> & WithStyle = React.memo(props => {
    const childRef = useRef(null),
        {
            addColumnMaxSize,
            config,
            data,
            rowId,
            isRowSelected,
            isRowClickDisabled,
            isRowSelectionDisabled,
            onRowSelection,
            dottedFieldName,
            isLoading
        } = props,
        isRowSelectionCell = useMemo(() => config.field === 'medly-table-checkbox', [config.field]);

    useEffect(() => {
        if (childRef.current) {
            const size = childRef.current.clientWidth;
            !isLoading && addColumnMaxSize(dottedFieldName, size);
        }
    }, [childRef.current]);

    const stopPropagation = useCallback((e: React.MouseEvent) => e.stopPropagation(), []),
        handleCellClick = useCallback((e: React.MouseEvent) => isRowSelectionCell && e.stopPropagation(), []),
        handleRowSelection = (id: number) => (e: React.FormEvent<HTMLInputElement>) => onRowSelection(id);

    const rowSelectionCheckbox = useMemo(
            () => (
                <Checkbox
                    disabled={isRowSelectionDisabled}
                    ref={childRef}
                    checked={isRowSelected}
                    onChange={handleRowSelection(rowId)}
                    onClick={stopPropagation}
                    name="active"
                />
            ),
            [rowId, isRowSelectionDisabled, isRowSelected]
        ),
        formatedCell = useCallback(() => {
            switch (config.formatter) {
                case 'boolean':
                    return (
                        <Text ref={childRef} textSize="M3">
                            {data ? 'Yes' : 'No'}
                        </Text>
                    );
                case 'react-component': {
                    const Component = config.component;
                    return (
                        <Styled.CustomComponentWrapper ref={childRef}>
                            <Component {...{ data, rowId, disabled: isRowClickDisabled }} />
                        </Styled.CustomComponentWrapper>
                    );
                }
                default:
                    return (
                        <Text ref={childRef} textSize="M3">
                            {data}
                        </Text>
                    );
            }
        }, [data, rowId, isLoading, isRowClickDisabled]);

    const textAlign = useMemo(() => config.align || (config.formatter === 'numeric' ? 'right' : 'left'), []);

    return (
        <Styled.Cell
            hide={config.hide}
            frozen={config.frozen}
            align={textAlign}
            onClick={handleCellClick}
            isRowSelectionCell={isRowSelectionCell}
        >
            {isLoading ? <Styled.LoadingDiv ref={childRef} /> : isRowSelectionCell ? rowSelectionCheckbox : formatedCell()}
        </Styled.Cell>
    );
});
Cell.displayName = 'Cell';
Cell.Style = Styled.Cell;

export default Cell;
