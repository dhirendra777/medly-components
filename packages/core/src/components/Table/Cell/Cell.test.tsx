import { render } from '@test-utils';
import React from 'react';
import Checkbox from '../../Checkbox';
import Text from '../../Text';
import { Cell } from './Cell.styled';

describe('Cell', () => {
    it('should render text properly', () => {
        const { container } = render(
            <Cell>
                <Text>Dummy</Text>
            </Cell>
        );
        expect(container).toMatchSnapshot();
    });

    it('should render frozen text properly', () => {
        const { container } = render(
            <Cell frozen>
                <Text>Dummy</Text>
            </Cell>
        );
        expect(container).toMatchSnapshot();
    });

    it('should not render text if hide property is given', () => {
        const { container } = render(
            <Cell hide>
                <Text>Dummy</Text>
            </Cell>
        );
        expect(container).toMatchSnapshot();
    });

    it('should render checkbox properly', () => {
        const { container } = render(
            <Cell hide>
                <Checkbox label="Dummy" />
            </Cell>
        );
        expect(container).toMatchSnapshot();
    });
});
