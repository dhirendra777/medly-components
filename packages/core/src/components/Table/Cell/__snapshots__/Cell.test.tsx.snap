// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell should not render text if hide property is given 1`] = `
.c2 {
  font-size: 14px;
  font-weight: 400;
  color: inherit;
}

.c3 .c1 {
  white-space: nowrap;
}

.c4 .c1 {
  white-space: nowrap;
}

.c0 {
  opacity: 0;
  padding: 0;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  overflow: hidden;
  border-right: 1px solid #E0E0E0;
}

.c0 .c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0px;
}

.c0 .c6 .c5 {
  margin-right: 0px;
}

.c0 .c1 {
  white-space: nowrap;
}

<div>
  <div
    class="c0"
  >
    <span
      class="c1 c2"
    >
      Dummy
    </span>
  </div>
</div>
`;

exports[`Cell should render checkbox properly 1`] = `
.c10 {
  font-size: 16px;
  cursor: pointer;
}

.c10 * {
  fill: #616161;
}

.c3 {
  font-size: 14px;
  font-weight: 400;
  color: #616161;
  margin: 0px 0px 0px 8px;
  cursor: pointer;
}

.c4 {
  grid-area: label;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c2 {
  margin: 8px 8px 8px 0px;
  display: inline-grid;
  grid-template-columns: max-content minmax(max-content,25%);
  grid-template-areas: 'field label' 'description random';
}

.c2 .c14 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
}

.c5 {
  position: relative;
  width: 16px;
  height: 16px;
  border: 1px solid #E0E0E0;
  border-radius: 15%;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c5:focus-within {
  box-shadow: 0 0 0 1pt #012040;
}

.c5 > * {
  width: 100%;
  height: 100%;
  border-radius: 15%;
}

.c8 {
  z-index: 1;
  position: absolute;
  background-color: #ffffff;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c8 .c9 {
  width: 0;
  height: 0;
  -webkit-transition: width 0.2s ease,height 0.2s ease;
  transition: width 0.2s ease,height 0.2s ease;
}

.c6 {
  opacity: 0;
  z-index: 2;
  position: absolute;
  margin: 0;
  cursor: pointer;
}

.c6:focus {
  outline: none;
}

.c6:disabled {
  cursor: not-allowed;
}

.c6:disabled + .c7 {
  background: #E0E0E0;
}

.c6:checked + .c7 .c9 {
  width: calc(70%);
  height: calc(70%);
}

.c11 .c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0px;
}

.c11 .c1 .c9 {
  margin-right: 0px;
}

.c12 .c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0px;
}

.c12 .c1 .c9 {
  margin-right: 0px;
}

.c0 {
  opacity: 0;
  padding: 0;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  overflow: hidden;
  border-right: 1px solid #E0E0E0;
}

.c0 .c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0px;
}

.c0 .c1 .c9 {
  margin-right: 0px;
}

.c0 .c13 {
  white-space: nowrap;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1 c2"
      id="Dummy-checkbox"
    >
      <label
        class="c3 c4"
        for="Dummy"
      >
        Dummy
      </label>
      <div
        class="c5"
      >
        <input
          class="c6"
          id="Dummy"
          type="checkbox"
        />
        <div
          class="c7 c8"
        >
          <svg
            class="c9 c10"
            height="1em"
            viewBox="0 0 18 14"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 11.17L1.83 7 .41 8.41 6 14 18 2 16.59.59z"
              fill="none"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Cell should render frozen text properly 1`] = `
.c2 {
  font-size: 14px;
  font-weight: 400;
  color: inherit;
}

.c3 .c1 {
  white-space: nowrap;
}

.c0 {
  opacity: 1;
  padding: 5px 10px;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  overflow: hidden;
  border-right: 1px solid #E0E0E0;
  position: -webkit-sticky;
  position: sticky;
  left: 0;
  z-index: 2;
  background-color: inherit;
}

.c0 .c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0px;
}

.c0 .c5 .c4 {
  margin-right: 0px;
}

.c0 .c1 {
  white-space: nowrap;
}

.c0 * {
  z-index: 2;
}

<div>
  <div
    class="c0"
  >
    <span
      class="c1 c2"
    >
      Dummy
    </span>
  </div>
</div>
`;

exports[`Cell should render text properly 1`] = `
.c2 {
  font-size: 14px;
  font-weight: 400;
  color: inherit;
}

.c0 {
  opacity: 1;
  padding: 5px 10px;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  overflow: hidden;
  border-right: 1px solid #E0E0E0;
}

.c0 .c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0px;
}

.c0 .c4 .c3 {
  margin-right: 0px;
}

.c0 .c1 {
  white-space: nowrap;
}

<div>
  <div
    class="c0"
  >
    <span
      class="c1 c2"
    >
      Dummy
    </span>
  </div>
</div>
`;
